File: problems/树/117.填充每个节点的下一个右侧节点指针 II.py
#!/usr/bin/env python3
"""
117. 填充每个节点的下一个右侧节点指针 II
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def connect(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root: return root
    dummy = TreeNode(0)
    cur = dummy
    while root:
        if root.left:
            cur.next = root.left
            cur = cur.next
        if root.right:
            cur.next = root.right
            cur = cur.next
        root = root.next
        if not root:
            cur = dummy
            root = dummy.next
            dummy.next = None
    return dummy

if __name__ == "__main__":
    root = TreeNode(1, TreeNode(2, None, TreeNode(4)), TreeNode(3, TreeNode(5), None))
    result = connect(root)
    print(result.val)  # 1
    print(connect(None))  # None

File: problems/树/117.填充每个节点的下一个右侧节点指针 II.cpp
#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode* next;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr), next(nullptr) {}
};

TreeNode* connect(TreeNode* root) {
    if (!root) return root;
    TreeNode* dummy = new TreeNode(0);
    TreeNode* cur = dummy;
    while (root) {
        if (root->left) { cur->next = root->left; cur = cur->next; }
        if (root->right) { cur->next = root->right; cur = cur->next; }
        root = root->next;
        if (!root) { cur = dummy; root = dummy->next; dummy->next = nullptr; }
    }
    return dummy;
}
int main() {
    TreeNode* root = new TreeNode(1);
    root->left  = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->right->left = new TreeNode(5);
    TreeNode* res = connect(root);
    cout << res->val << endl;  // 1
    cout << (connect(nullptr) == nullptr ? "null" : "?") << endl;  // null
    return 0;
}
